### PR TITLE
Don't do proxy selection eagerly if we're pooling

### PR DESCRIPTION
**Pull request by swankjesse**
If we're going to end up using a pooled connection anyway we
should defer proxy selection until we need it.

### CODE REVIEW

File: okhttp/src/test/java/okhttp3/EventListenerTest.java [hunk 1]
```diff
@@ -1258,9 +1254,7 @@ private void timeToFirstByte() throws IOException {
 
     // Confirm the events occur when expected.
     listener.takeEvent(CallStart.class, 0L);
-    listener.takeEvent(ProxySelectStart.class, applicationInterceptorDelay);
-    listener.takeEvent(ProxySelectEnd.class, 0L);
-    listener.takeEvent(ConnectionAcquired.class, 0L);
+    listener.takeEvent(ConnectionAcquired.class, applicationInterceptorDelay);
```
Comment (yschimke) [inline okhttp/src/test/java/okhttp3/EventListenerTest.java]: That's a lovely change.

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt [hunk 2]
```diff
@@ -690,7 +690,7 @@ class RealConnection(
       if (e is StreamResetException) {
         when (e.errorCode) {
           ErrorCode.REFUSED_STREAM -> {
-            // Retry REFUSED_STREAM errors once on the same connection.
+            // Stop using this connection after one REFUSED_STREAM error.
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt]: two?

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 3]
```diff
@@ -287,12 +291,21 @@ class ExchangeFinder(
...
         nextRouteToTry = call.connection!!.route()
         return true
       }
-      return (routeSelection?.hasNext() ?: false) || routeSelector.hasNext()
+
+      // If we have a routes left, use 'em.
+      if (routeSelection?.hasNext() == true) return true
+
+      // If we haven't initialized the route selector yet, assume it'll have at least one route.
+      val localRouteSelector = routeSelector ?: return true
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: Are these the lines that defer proxy lookup?

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt]: Yeah, that doesn’t read as well as I wanted. Fixed:

```
            // Stop using this connection on the 2nd REFUSED_STREAM error.
```

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: You got it.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: Nope, I read this wrong. Deferring the proxy lookup happens by deferring RouteSelector construction.

File: okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt [hunk 4]
```diff
@@ -57,8 +57,7 @@ class ExchangeFinder(
   private var routeSelection: RouteSelector.Selection? = null
 
   // State guarded by connectionPool.
-  private val routeSelector: RouteSelector = RouteSelector(
-      address, call.client.routeDatabase, call, eventListener)
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt]: Deferring proxy selection happens here.

**Comment by yschimke** [comment]
@swankjesse  I assume if we don't find exact match then we still do the DNS check before reusing a pooled connection via connection coalescing rules?

**Comment by swankjesse** [comment]
I got the branch name wrong because we were already deferring DNS. Here's the new sequence:

 1. Look in the pool
 2. Do proxy selection and the first DNS lookup
 3. Look in the pool again. This might hit because we have an IP address
 4. Connect to the resolved IP address
 5. Look in the pool the last time to see if we raced

It's a bit obnoxious that we look in the pool 3 times for unpooled connections! But this is how we maximize the pool’s use while minimizing network activity.

**Pull request merged**